cpp/src/parquet/chunker_internal.cc

```diff
@@ -0,0 +1,413 @@
+      has_matched_ = has_matched_ || ((rolling_hash_ & rolling_hash_mask_) == 0);
+    }
+  }
+
+  bool NeedNewChunk() {
+    // decide whether to create a new chunk based on the rolling hash; has_matched_ is
+    // set to true if we encountered a match since the last NeedNewChunk() call
+    if (ARROW_PREDICT_FALSE(has_matched_)) {
+      has_matched_ = false;
+      // in order to have a normal distribution of chunk sizes, we only create a new chunk
+      // if the adjused mask matches the rolling hash 8 times in a row, each run uses a
+      // different gearhash table (gearhash's chunk size has geometric distribution, and
+      // we use central limit theorem to approximate normal distribution, see
+      // section 6.2.1 in paper https://www.cidrdb.org/cidr2023/papers/p43-low.pdf)
+      if (ARROW_PREDICT_FALSE(++nth_run_ >= kNumGearhashTables)) {
+        // note that we choose not to reset the rolling hash state here, nor anywhere else
+        // in the code, in practice this doesn't seem to affect the chunking effectiveness
+        nth_run_ = 0;
+        chunk_size_ = 0;
+        return true;
+      }
+    }
+    if (ARROW_PREDICT_FALSE(chunk_size_ >= max_chunk_size_)) {
+      // we have a hard limit on the maximum chunk size, note that we don't reset the
+      // rolling hash state here, so the next NeedNewChunk() call will continue from the
+      // current state
+      chunk_size_ = 0;
+      return true;
+    }
+    return false;
+  }
+
+  void ValidateChunks(const std::vector<Chunk>& chunks, int64_t num_levels) const {
+    // chunks must be non-empty and monotonic increasing
+    ARROW_DCHECK(!chunks.empty());
+
+    // the first chunk must start at the first level
+    auto first_chunk = chunks.front();
+    ARROW_DCHECK_EQ(first_chunk.level_offset, 0);
+    ARROW_DCHECK_EQ(first_chunk.value_offset, 0);
+
+    // the following chunks must be contiguous, non-overlapping and monotonically
+    // increasing
+    auto sum_levels = first_chunk.levels_to_write;
+    for (size_t i = 1; i < chunks.size(); ++i) {
+      auto chunk = chunks[i];
+      auto prev_chunk = chunks[i - 1];
+      ARROW_DCHECK_GT(chunk.levels_to_write, 0);
+      ARROW_DCHECK_GE(chunk.value_offset, prev_chunk.value_offset);
+      ARROW_DCHECK_EQ(chunk.level_offset,
+                      prev_chunk.level_offset + prev_chunk.levels_to_write);
+      sum_levels += chunk.levels_to_write;
+    }
+    ARROW_DCHECK_EQ(sum_levels, num_levels);
+
+    // the last chunk must end at the last level
+    auto last_chunk = chunks.back();
+    ARROW_DCHECK_EQ(last_chunk.level_offset + last_chunk.levels_to_write, num_levels);
+  }
+
+  template <typename RollFunc>
+  std::vector<Chunk> Calculate(const int16_t* def_levels, const int16_t* rep_levels,
+                               int64_t num_levels, const RollFunc& RollValue) {
+    // Calculate the chunk boundaries for typed Arrow arrays.
```

Comment:
I don't remember if we discussed it already, but is there a reason we don't reset `nth_run_`, `chunk_size_` and `rolling_hash_` here?

Either way, we should add a comment explaining why.

Reply:
The chunking process applies to the entire column across all pages and row groups and only initialized when the column writer is created. This way chunking can be continued between different WriteArrow calls.

Reply:
Added a more detailed description to the method.